README
Support pgbouncer 1.7.X and 1.8.X

Extended from http://git.cbaines.net/prometheus-pgbouncer-exporter/

  Add re-connect once when pgbouncer connection is closed.
  Add parameter pgbouncer-password for support pgbouncer password.
  Add version check and support PgBouncer 1.8.

# requirements
ConfigArgParse
prometheus_client
psycopg2

# Usage:
python3 setup.py install
/usr/bin/prometheus-pgbouncer-exporter --pgbouncer-host=127.0.0.1 --pgbouncer-port=6432 --pgbouncer-user=stats --pgbouncer-password=your_password

=======================================================================================

# PgBouncer exporter for Prometheus

This is a simple exporter for PgBouncer that makes several metrics available to
Prometheus.

Metrics are exported from the SHOW LISTS, STATS, POOLS and DATABASES command
output. For the full list, see the
prometheus\_pgbouncer\_exporter/collectors.py file.

Please send any comments or queries to Christopher Baines <mail@cbaines.net>.

## Installation

### Debian

You can build a Debian binary package from the source package in the git
repository. Alternatively, you can download a binary package from here:

<http://cbaines.net/projects/prometheus-pgbouncer-exporter/releases/>

### Other Operating Systems

I would recommend installing using setuptools, then running the
promtheus-pgbouncer-exporter script. A systemd service file is provided which
can be used if you have systemd.

## Authentication

The service connects to the pgbouncer admin console to gather metrics. The
service file runs the service as the postgres user (which is assumed to be the
user which pgbouncer is running as), such that it can access the admin console
(for which access is allowed if the login comes from via a Unix socket and the
client has the same user id as the pgbouncer service).

This setup does mean that the exporter service (when running as the postgres
user) has far more capabilities than it requires. A more secure approach is to
run the service as a unprivileged user, which is listed in the stats_users
configuration parameter, as this means the process does not have to run as the
postgres user, and will be restricted to using the SHOW command.


CHANGELOG
prometheus-pgbouncer-exporter (1.8)

  * Include the port setting in the example configuration file
  * Add note in the README about authentication

 -- petanne <lywxj1992@gmail.com>  Fri, 12 Jan 2018 13:44:50 +0000

prometheus-pgbouncer-exporter (1.7)

  * Include the port setting in the example configuration file
  * Add note in the README about authentication

 -- Christopher Baines <mail@cbaines.net>  Tue, 01 Mar 2016 08:45:45 +0000

prometheus-pgbouncer-exporter (1.6)

  * Make the pgbouncer host configurable
  * Fix help strings

 -- Christopher Baines <mail@cbaines.net>  Sat, 27 Feb 2016 17:55:20 +0000

prometheus-pgbouncer-exporter (1.5)

  * Make the port and host configurable
  * Remove default config file path
  * Remove debian/
  * Separate changelog
  * Use add --pgbouncer to port and user options

 -- Christopher Baines <mail@cbaines.net>  Sat, 27 Feb 2016 13:37:52 +0000

prometheus-pgbouncer-exporter (1.4)

  * New release
   - Add manpage

 -- Christopher Baines <mail@cbaines.net>  Wed, 17 Feb 2016 11:46:16 +0000

prometheus-pgbouncer-exporter (1.3)

  * New release
   - Readme improvements

 -- Christopher Baines <mail@cbaines.net>  Mon, 15 Feb 2016 16:44:28 +0000

prometheus-pgbouncer-exporter (1.2)

  * New release
    - Add extra documentation
    - Add extra licence information
    - Add --version argument
    - Rename metrics, removing _count prefix

 -- Christopher Baines <mail@cbaines.net>  Sat, 16 Jan 2016 20:47:22 +0000

prometheus-pgbouncer-exporter (1.1)

  * New release
    - Improve the .service file
    - Fix a metric name

 -- Christopher Baines <mail@cbaines.net>  Tue, 05 Jan 2016 10:05:06 +0000

prometheus-pgbouncer-exporter (1.0)

  * Initial release

 -- Christopher Baines <mail@cbaines.net>  Sun, 27 Dec 2015 14:57:54 +0000


tests/test_connection.py
import unittest

from unittest.mock import patch, Mock

from prometheus_pgbouncer_exporter import utils

class ConnectionTest(unittest.TestCase):

    @patch('prometheus_pgbouncer_exporter.utils.psycopg2.connect')
    def test_get_connection_with_host_and_dbname(self, connect):
        conn = utils.get_connection(host='/tmp/', dbname='template1')
        connect.assert_called_once_with(host='/tmp/', dbname='template1')
        connect.return_value.set_session.assert_called_once_with(autocommit=True)

    @patch('prometheus_pgbouncer_exporter.utils.psycopg2.connect')
    def test_get_connection_with_old_args_plus_password_works(self, connect):
        conn = utils.get_connection(user='pablo', port=5432, host='localhost', password='mypassword')
        connect.assert_called_once_with(dbname='pgbouncer', user='pablo', port=5432, host='localhost', password='mypassword')


prometheus_pgbouncer_exporter/utils.py
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import logging
import psycopg2
import os
import re


def get_connection(user, password, port, host):
    connection = psycopg2.connect(
        database='pgbouncer',
        user=user,
        password=password,
        port=port,
        host=host
    )

    # pgbouncer does not support transactions (as it does not make sense to),
    # so don't start a transaction when connecting
    connection.set_session(autocommit=True)

    return connection


def get_data_by_named_column(connection, key):
    with connection.cursor() as cursor:
        cursor.execute('SHOW %s;' % key)

        rows = cursor.fetchall()
        column_names = list(column.name for column in cursor.description)

    return [
        dict(zip(column_names, row))
        for row in rows
    ]


def get_data_by_named_row(connection, key):
    with connection.cursor() as cursor:
        cursor.execute('SHOW %s;' % key)

        return dict(cursor.fetchall())


def get_pgbouncer_version():
    version = '1.8'
    s = os.popen('pgbouncer --version').read()
    res = re.match('.*version (?P<v1>\d+)\.(?P<v2>\d+)\..*', s).groupdict()
    try:
        if int(res['v1']) <= 1 and int(res['v2']) <= 7:
            version = '1.7'
    except:
        logging.warning('get bgbouncer version failed, default to 1.8')
    return version


prometheus_pgbouncer_exporter/collectors.py
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from prometheus_client.core import GaugeMetricFamily

from .utils import get_data_by_named_column, get_data_by_named_row
from .conn import CONN


class PgBouncerCollector(object):
    def __init__(self, connection=None, namespace='pgbouncer'):
        self.connection = connection
        self.namespace = namespace


class NamedColumnCollector(PgBouncerCollector):
    def get_labels_for_row(self, row):
        raise NotImplementedError()

    def collect(self):
        # Each of the metrics is recorded per database, by calling add_metric
        # with the database as the label on the appropriate GaugeMetricFamily.

        # First, create the GaugeMetricFamily objects for each metric
        gauges = {
            key: GaugeMetricFamily(
                name="%s_%s" % (self.namespace, name),
                documentation="%s (%s)" % (documentation, key),
                labels=self.labels,
            )
            for key, (name, documentation) in self.metrics.items()
        }

        rows = get_data_by_named_column(CONN.CONNECTION, self.query_parameter)

        # Each row coresponds to the metrics for a particular database
        for row in rows:
            database = row['database']

            if self.databases is not None and database not in self.databases:
                continue

            # Sort for deterministic ordering in the output
            for key in sorted(self.metrics.keys()):
                value = row[key]
                label_values = self.get_labels_for_row(row)

                gauges[key].add_metric(label_values, value)

        for gauge in gauges.values():
            yield gauge


class ListsCollector(PgBouncerCollector):
    metrics = {
        'databases': (
            'databases',
            "Number of databases",
        ),
        'users': (
            'users',
            "Number of users",
        ),
        'pools': (
            'pools',
            "Number of pools",
        ),
        'free_clients': (
            'free_clients',
            "Number of free clients",
        ),
        'used_clients': (
            'used_clients',
            "Number of used clients",
        ),
        'login_clients': (
            'login_clients',
            "Number of clients in the login stats",
        ),
        'free_servers': (
            'free_servers',
            "Number of free servers",
        ),
        'used_servers': (
            'used_servers',
            "Number of used servers",
        ),
        'dns_names': (
            'dns_names',
            "",
        ),
        'dns_zones': (
            'dns_zones',
            "",
        ),
        'dns_queries': (
            'dns_queries',
            "",
        ),
        'dns_pending': (
            'dns_pending',
            "",
        ),
    }

    def collect(self):
        data = get_data_by_named_row(CONN.CONNECTION, 'LISTS')

        for key, (name, documentation) in sorted(
                self.metrics.items(), key=lambda x: x[0]
        ):
            yield GaugeMetricFamily(
                name="%s_%s" % (self.namespace, name),
                documentation="%s (%s)" % (documentation, key),
                value=data[key],
            )


class StatsCollector(NamedColumnCollector):
    query_parameter = 'STATS'

    labels = ['database']

    metrics = {
        'total_requests': (
            'requests_total',
            "Total number of SQL requests pooled by pgbouncer",
        ),
        'total_received': (
            'received_bytes_total',
            "Total volume in bytes of network traffic received by pgbouncer",
        ),
        'total_sent': (
            'sent_bytes_total',
            "Total volume in bytes of network traffic sent by pgbouncer",
        ),
        'total_query_time': (
            'query_microseconds_total',
            "Total number of microseconds spent by pgbouncer when actively connected to PostgreSQL",
        ),
        'avg_query': (
            'avg_query',
            "Avg number of avg_query",
        ),
    }

    def __init__(self, databases, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.databases = databases

    def get_labels_for_row(self, row):
        return [row['database']]


class Stats18Collector(NamedColumnCollector):
    query_parameter = 'STATS'

    labels = ['database']

    metrics = {
        'total_query_count': (
            'requests_total',
            "Total number of SQL requests pooled by pgbouncer",
        ),
        'total_received': (
            'received_bytes_total',
            "Total volume in bytes of network traffic received by pgbouncer",
        ),
        'total_sent': (
            'sent_bytes_total',
            "Total volume in bytes of network traffic sent by pgbouncer",
        ),
        'total_query_time': (
            'query_microseconds_total',
            "Total number of microseconds spent by pgbouncer when actively connected to PostgreSQL",
        ),
        'avg_query_count': (
            'avg_query',
            "Avg number of avg_query",
        ),
    }

    def __init__(self, databases, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.databases = databases

    def get_labels_for_row(self, row):
        return [row['database']]


class PoolsCollector(NamedColumnCollector):
    query_parameter = 'POOLS'

    labels = ['database', 'user']

    metrics = {
        'cl_active': (
            'active_clients',
            "Client connections that are linked to server connection and can process queries",
        ),
        'cl_waiting': (
            'waiting_clients',
            "Client connections have sent queries but have not yet got a server connection",
        ),
    }

    def __init__(self, databases, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.databases = databases

    def get_labels_for_row(self, row):
        return [row['database'], row['user']]


class DatabasesCollector(NamedColumnCollector):
    query_parameter = 'DATABASES'

    labels = ['database']

    metrics = {
        'pool_size': (
            'pool_size',
            "",
        ),
        'reserve_pool': (
            'reserve_pool',
            "",
        ),
    }

    def __init__(self, databases, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.databases = databases

    def get_labels_for_row(self, row):
        return [row['database']]


prometheus_pgbouncer_exporter/exposition.py
# Copyright (C) 2015  Christopher Baines <mail@cbaines.net>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
import psycopg2

from prometheus_client.exposition import MetricsHandler, generate_latest, CONTENT_TYPE_LATEST
from prometheus_client import core
from urllib.parse import quote_plus, parse_qs, urlparse

from .conn import CONN

index_page = """
<!doctype html>

<html lang="en">
<head>
  <meta charset="utf-8">

  <title>Prometheus PgBouncer Exporter</title>
</head>

<body>
  <h1>PgBouncer exporter for Prometheus</h1>

  <p>
    This is a simple exporter for PgBouncer that makes several metrics
    available to Prometheus.
  </p>

  <p>
    Metrics are exported from the SHOW LISTS, STATS, POOLS and DATABASES comand
    output.
  </p>

  <a href="/debug/metrics">View metrics</a>

  <h2>Example Configuration</h2>
  <p>
    You must configure Prometheus to scrape the metrics exported here. The port
    is 9127, and the configuration should look something like the example
    below.
  </p>
  <pre><code>
    scrape_configs:
      - job_name: pgbouncer
        target_groups:
          - targets:
            - MACHINE_ADDRESS:9127
  </code></pre>

  <h2>Information</h2>
  <p>
    Copyright (C) 2015  Christopher Baines <mail@cbaines.net><br>
    <a href="/licence">View Licence</a>
  </p>

  <p>
    The source may be obtained from
    <a href="http://git.cbaines.net/prometheus-pgbouncer-exporter/">
    this Git repository</a>
  </p>

  <p>
    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.
  </p>

  <p>
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.
  </p>
</body>
</html>
"""


def create_request_handler(licence_location):
    class RequestHandler(MetricsHandler):
        def do_GET_super(self):
            """extended super MetricsHandler().do_GET()
                add reconnect when psycopg2.Error
            """
            registry = core.REGISTRY
            params = parse_qs(urlparse(self.path).query)
            if 'name[]' in params:
                registry = registry.restricted_registry(params['name[]'])
            try:
                output = generate_latest(registry)
            except psycopg2.Error:
                CONN.reconnect()
                # generate_latest again after re-connect
                try:
                    output = generate_latest(registry)
                except:
                    self.send_error(500, 'error generating metric output')
                    raise
            self.send_response(200)
            self.send_header('Content-Type', CONTENT_TYPE_LATEST)
            self.end_headers()
            self.wfile.write(output)

        def do_GET(self):
            if self.path == "/debug/metrics":
                return self.do_GET_super()

            self.send_response(200)
            self.end_headers()

            if self.path == "/licence":
                with open(
                        licence_location,
                        'rb',
                ) as licence:
                    self.wfile.write(licence.read())
            else:
                self.wfile.write(index_page.encode('UTF-8'))

    return RequestHandler


prometheus_pgbouncer_exporter/cli.py
#!/usr/bin/python3

# Copyright (C) 2015  Christopher Baines <mail@cbaines.net>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

# 2017年10月19日
# 开发日志：
# 解决pgbouncer重启时，pgbouncer_exporter与pgbouncer的长连接失效且无法重新连接的问题
# 增加全局变量 from .conn import CONN
# 增加了CONN的reconnect()重连方法
# 增加支持pgbouncer password参数

import logging
import configargparse

from os.path import join, dirname, normpath
from http.server import HTTPServer
from prometheus_client.core import REGISTRY

from . import __version__
from .utils import get_connection, get_pgbouncer_version
from .exposition import create_request_handler
from .collectors import StatsCollector, Stats18Collector, ListsCollector, PoolsCollector, \
    DatabasesCollector

from .conn import CONN


def main():
    p = configargparse.ArgParser(
        prog="prometheus-pgbouncer-exporter",
    )

    p.add(
        '--version',
        action='store_true',
        help="Show the version",
    )

    p.add(
        '-c',
        '--config',
        is_config_file=True,
        help='config file path',
        env_var='CONFIG',
    )

    p.add(
        '--port',
        default='9127',
        help="Port on which to expose metrics",
        type=int,
        env_var='PORT',
    )
    p.add(
        '--host',
        default='0.0.0.0',
        help="Host on which to expose metrics",
        env_var='HOST',
    )

    p.add(
        '--pgbouncer-port',
        default='6432',
        help="Port to connect to pgbouncer",
        env_var='PGBOUNCER_PORT',
    )
    p.add(
        '--pgbouncer-user',
        default='pgbouncer',
        help="User to connect to pgbouncer with",
        env_var='PGBOUNCER_USER',
    )
    p.add(
        '--pgbouncer-password',
        default='your_password',
        help="Password to connect to pgbouncer with",
        env_var='PGBOUNCER_PASSWORD',
    )
    p.add(
        '--pgbouncer-host',
        default=None,
        help="Host on which to connect to pgbouncer",
        env_var='PGBOUNCER_HOST',
    )

    p.add(
        '--database',
        action='append',
        help="Databases to report metrics for, if this is not specified, all metrics will be reported",
        env_var='PGBOUNCER_DATABASES',
    )

    p.add(
        '--licence-location',
        default=join(dirname(dirname(normpath(__file__))), 'LICENSE'),
        help="The location of the licence, linked to through the web interface",
        env_var='LICENCE_LOCATION',
    )

    options = p.parse_args()

    if options.version:
        print("prometheus-pgbouncer-exporter %s" % __version__)
        return

    logging.basicConfig(level=logging.DEBUG)

    logging.info(p.format_values())

    connection = get_connection(
        options.pgbouncer_user,
        options.pgbouncer_password,
        options.pgbouncer_port,
        options.pgbouncer_host,
    )
    # global CONNECTION
    CONN.CONNECTION = connection
    CONN.PGBOUNCER_USER = options.pgbouncer_user
    CONN.PGBOUNCER_PASSWORD = options.pgbouncer_user
    CONN.PGBOUNCER_PORT = options.pgbouncer_port
    CONN.PGBOUNCER_HOST = options.pgbouncer_host

    version = get_pgbouncer_version()
    if version == '1.8':
        REGISTRY.register(Stats18Collector(
            databases=options.database,
        ))
    else:
        REGISTRY.register(StatsCollector(
            databases=options.database,
        ))

    REGISTRY.register(PoolsCollector(
        databases=options.database,
    ))

    REGISTRY.register(DatabasesCollector(
        databases=options.database,
    ))

    REGISTRY.register(ListsCollector(
    ))

    httpd = HTTPServer(
        (options.host, options.port),
        create_request_handler(options.licence_location),
    )

    logging.info("Listing on port %s:%d" % (options.host, options.port))

    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        pass
    finally:
        connection.close()
